src/Effect/ReadFS.hs

```diff
@@ -168,34 +151,31 @@ readContentsXML file = do
     Right a -> pure a
 
 -- | Read XML from a file
-readContentsXML' :: (FromXML a, Has ReadFS sig m, Effect sig) => Path b File -> m (Either ReadFSErr a)
+readContentsXML' :: (FromXML a, Has ReadFS sig m) => Path b File -> m (Either ReadFSErr a)
 readContentsXML' = runError . readContentsXML
 
-runReadFSIO :: ReadFSIOC m a -> m a
-runReadFSIO = coerce
-
-newtype ReadFSIOC m a = ReadFSIOC { runReadFSIOC :: m a }
+newtype ReadFSIOC m a = ReadFSIOC {runReadFSIO :: m a}
   deriving (Functor, Applicative, Monad, MonadIO, MonadFail)
 
 instance (Algebra sig m, MonadIO m) => Algebra (ReadFS :+: sig) (ReadFSIOC m) where
-  alg (L act) = case act of
-    ReadContentsBS' file k -> (k =<<) . ReadFSIOC $ liftIO $
-      (Right <$> BS.readFile (toFilePath file))
-      `E.catch`
-      (\(e :: E.IOException) -> pure (Left (FileReadError (toFilePath file) (T.pack (show e)))))
-    ReadContentsText' file k -> (k =<<) . ReadFSIOC $ liftIO $
-      (Right . decodeUtf8 <$> BS.readFile (toFilePath file))
-      `E.catch`
-      (\(e :: E.IOException) -> pure (Left (FileReadError (toFilePath file) (T.pack (show e)))))
-    ResolveFile' dir path k -> (k =<<) . ReadFSIOC $ liftIO $
-      (Right <$> PIO.resolveFile dir (T.unpack path))
-      `E.catch`
-      (\(e :: E.IOException) -> pure (Left (ResolveError (toFilePath dir) (T.unpack path) (T.pack (show e)))))
-    ResolveDir' dir path k -> (k =<<) . ReadFSIOC $ liftIO $
-      (Right <$> PIO.resolveDir dir (T.unpack path))
-      `E.catch`
-      (\(e :: E.IOException) -> pure (Left (ResolveError (toFilePath dir) (T.unpack path) (T.pack (show e)))))
-    -- NB: these never throw
-    DoesFileExist file k -> k =<< PIO.doesFileExist file
-    DoesDirExist dir k -> k =<< PIO.doesDirExist dir
-  alg (R other) = ReadFSIOC (alg (handleCoercible other))
+  alg hdl sig ctx = ReadFSIOC $ do
```

Comment:
Algebra instances are still complex if you're not familiar with how fused-effects works. I'll need to write some documentation for our team to build intuition here